designs/no-schedule-taint.md

```diff
@@ -0,0 +1,37 @@
+
+Karpenter should taint nodes with a `karpenter.sh/disrupting:NoSchedule` taint rather than relying on the upstream unschedulable taint.
+
+Since Karpenter is currently migrating to v1beta1, this behavior change will only be present with v1beta1 APIs.
+
+## Considerations
+
+### Daemonsets
+
+The [Daemonset Controller](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#create-a-daemonset) adds [default tolerations to every pod](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#taints-and-tolerations), including the `node.kubernetes.io/unschedulable` taint. As part of termination, Karpenter [will not](https://github.com/aws/karpenter-core/blob/main/pkg/controllers/termination/terminator/terminator.go#L79-L81) evict pods that tolerate the `node.kubernetes.io/unschedulable` taint, as they would immediately reschedule. After this change, Karpenter will not use the `node.kubernetes.io/unschedulable` taint, resulting in draining for daemonsets during termination. If users don’t want their daemonsets to be evicted by Karpenter at termination, they’ll need to add a special toleration, not currently added by default. **This will be called out in the v1beta1 migration.**
+
+Cluster Autoscaler [uses their own `NoSchedule` taint](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/utils/taints/taints.go#L39-L42) to ensure that pods will not schedule to the nodes being deprovisioned. [By default](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/main.go#L216-L217), CAS will evict daemonset pods for non-empty nodes and will also respect a pod [annotation](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-can-i-enabledisable-eviction-for-a-specific-daemonset) on each daemonset pod to know if Karpenter should evict the pod.
+
+🔑 Karpenter determines if it should evict a pod through its tolerations. Karpenter could consider adding in a configuration surface similar to CAS to allow evicting pods owned by a daemonset. I recommend that we make this taint change during v1beta1 with ample documentation, and add the configuration surface in as a follow up if requested.
+
+### SIG-Autoscaling Alignment
+
+Karpenter is currently in process of aligning with SIG-Autoscaling for a common experience. As part of alignment, all API groups for both projects are proposed to be renamed to `node-lifecycle.kubernetes.io`.  After this change, users migrating to `v1beta1` will need to change their daemonsets to tolerate the new Karpenter taint if they want their daemonsets not evicted. If the API group is changed as part of alignment, these API groups will change once more, requiring users to change their tolerations on their workloads once again, introducing more churn. As this alignment will be a breaking change for both CAS and Karpenter, this change is inevitable, and should be done when Karpenter once again at a future time, potentially when graduating APIs to `v1`.
```

Comment:
I'm not sure what I think about the last statement, "this change is inevitable". I don't think that this is accurate. I think the real truth here is there is a lot of ambiguity around the direction here. I'd say something like, "There is currently a lot of ambiguity on the shared direction of Karpenter, CAS, and the entirety of upstream Kubernetes aligning around these concepts. It's unclear how this conversation will evolve and how long this conversation will take to resolve. As a result, we are choosing not to block on this conversation and, instead, taking an opinionated stance in Karpenter that we can choose to evolve later."

Reply:
Good call! thanks.